src/pages/Compare/index.tsx

```diff
@@ -40,13 +41,25 @@ export class CompareBase extends React.Component<Props> {
   };
 
   componentDidMount() {
-    const { _fetchVersion, dispatch, match } = this.props;
-    const { addonId, baseVersionId } = match.params;
+    const { _fetchVersion, dispatch, history, match } = this.props;
+    const { lang, addonId, baseVersionId, headVersionId } = match.params;
+
+    let oldVersionId = parseInt(baseVersionId, 10);
+    let newVersionId = parseInt(headVersionId, 10);
+
+    // We ensure the new version ID is newer than the old version ID.
+    if (oldVersionId > newVersionId) {
+      history.push(
+        `/${lang}/compare/${addonId}/versions/${headVersionId}...${baseVersionId}/`,
+      );
+      oldVersionId = newVersionId;
+      newVersionId = parseInt(baseVersionId, 10);
+    }
 
     dispatch(
       _fetchVersion({
         addonId: parseInt(addonId, 10),
-        versionId: parseInt(baseVersionId, 10),
+        versionId: oldVersionId,
```

Comment:
This is a convenient way to do it but it's confusing because the component is no longer fetching the version requested in the URL. This makes the logic hard to follow and is going to make it hard to debug later on. 

To keep the logic nice and simple, we need to migrate to a `useEffect` hook but until then I would like to see the fetching logic changed to always use the URL parameter. I know it's a bit more complicated and will require more tests but this is more like what we need:

```diff
diff --git a/src/pages/Compare/index.tsx b/src/pages/Compare/index.tsx
index 7f8bc7f..764cca7 100644
--- a/src/pages/Compare/index.tsx
+++ b/src/pages/Compare/index.tsx
@@ -52,16 +52,33 @@ export class CompareBase extends React.Component<Props> {
       history.push(
         `/${lang}/compare/${addonId}/versions/${headVersionId}...${baseVersionId}/`,
       );
-      oldVersionId = newVersionId;
-      newVersionId = parseInt(baseVersionId, 10);
+      return;
     }
 
-    dispatch(
-      _fetchVersion({
-        addonId: parseInt(addonId, 10),
-        versionId: oldVersionId,
-      }),
-    );
+    this.loadData();
+  }
+
+  componentDidUpdate(oldProps: Props) {
+    this.loadData(oldProps);
+  }
+
+  loadData(oldProps?: Props) {
+    const { _fetchVersion, dispatch, match } = this.props;
+    const { addonId, baseVersionId } = match.params;
+    let oldVersionId = parseInt(baseVersionId, 10);
+
+    if (
+      !oldProps ||
+      addonId !== oldProps.match.params.addonId ||
+      baseVersionId !== oldProps.match.params.baseVersionId
+    ) {
+      dispatch(
+        _fetchVersion({
+          addonId: parseInt(addonId, 10),
+          versionId: oldVersionId,
+        }),
+      );
+    }
   }
 
   onSelectFile = () => {};
```

Reply:
I wasn't sure because these url params won't be used in the rest of this component and because there is a re-render, the children will receive the versions IDs that match the url (given that the url gets updated directly here) 0

Reply:
I'm not sure what you mean. The router prop `props.match.headVersionId` is passed to `_fetchVersion()` but your patch overloads that behavior. Why not always use the router prop and let `history.push()` be the one to change it? That's the patch I proposed.

Reply:
Sorry, let me rephrase: I was not sure implementing both `componentDidMount()` and `componentDidUpdate()` was useful because:

- the `baseVersionId` and `headVersionId` are only used in `componentDidMount()`, no need to invert them in `render()` for example (so the confusing behavior is small and scoped)
- as soon as `history.push()` is called, the URL is correct
- the `if` statement inverts the version IDs so that it matches the pushed URL, reflecting the URL seen by the user
- when children are rendered, they will receive the correct values in `baseVersionId` and `headVersionId` (if the children access the router props with `withRouter()`for example)

It's probably a good idea to early return after `history.push()` but this makes me cry already:

```
+    if (
+      !oldProps ||
+      addonId !== oldProps.match.params.addonId ||
+      baseVersionId !== oldProps.match.params.baseVersionId
+    ) {
```